tests/h/activity/query_test.py

```diff
@@ -379,6 +379,21 @@ def test_it_returns_each_annotations_group(self,
             else:
                 assert False
 
+    def test_it_returns_each_annotations_incontext_link(self,
+                                                  annotations,
+                                                  links,
+                                                  pyramid_request):
+        result = execute(pyramid_request, MultiDict(), self.PAGE_SIZE)
+
+        presented_annotations = []
+        for timeframe in result.timeframes:
+            for bucket in timeframe.document_buckets.values():
+                presented_annotations.extend(bucket.presented_annotations)
+
+        for annotation in annotations:
+            for presented_annotation in presented_annotations:
+                assert presented_annotation['incontext_link'] == links.incontext_link.return_value
```

Comment:
It's an awkward test to write but here it might be better to test that 1. `incontext_link()` is called once for each annotation, and 2. Each `presented_annotation['incontext_link']` contains the value that `incontext_link()` returned for that annotation. One way to do that is to replace `links.incontext_link()` with a fake function that, when called with a given annotation, returns a value that's predictable based on the annotation that was given.

This can be done using [mock's side_effect property](http://www.voidspace.org.uk/python/mock/mock.html#mock.Mock.side_effect) which allows you to nominate a function to be called when a mock is called, and the mock will return whatever that functions returns.

Something like this:

```diff
diff --git a/tests/h/activity/query_test.py b/tests/h/activity/query_test.py
index 0e7a085..2fbb7b6 100644
--- a/tests/h/activity/query_test.py
+++ b/tests/h/activity/query_test.py
@@ -383,6 +383,13 @@ class TestExecute(object):
                                                   annotations,
                                                   links,
                                                   pyramid_request):
+        def incontext_link(request, annotation):
+            assert request == pyramid_request, (
+                "It should always pass the request to incontext_link")
+            # Return a predictable per-annotation value for the incontext link.
+            return 'incontext_link_for_annotation_{id}'.format(id=id)
+        links.incontext_link.side_effect = incontext_link
+
         result = execute(pyramid_request, MultiDict(), self.PAGE_SIZE)
 
         presented_annotations = []
@@ -392,7 +399,8 @@ class TestExecute(object):
 
         for annotation in annotations:
             for presented_annotation in presented_annotations:
-                assert presented_annotation['incontext_link'] == links.incontext_link.return_value
+                assert presented_annotation['incontext_link'] == (
+                    'incontext_link_for_annotation_{id}'.format(id=id))
 
     def test_it_returns_the_total_number_of_matching_annotations(
             self, pyramid_request):
```